mempool/src/pool/reorg.rs

```diff
@@ -135,14 +141,31 @@ pub fn handle_new_tip<M: MemoryUsageEstimator>(
 ) -> Result<(), ReorgError> {
     mempool.rolling_fee_rate.get_mut().set_block_since_last_rolling_fee_bump(true);
 
-    let is_ibd = mempool.blocking_chainstate_handle().call(|cs| cs.is_initial_block_download())?;
-    if is_ibd {
-        log::debug!("Not updating mempool tx verifier during IBD");
+    let (is_ibd, actual_tip) = mempool.blocking_chainstate_handle().call(|cs| {
+        let is_ibd = cs.is_initial_block_download();
+        let actual_tip = cs.get_best_block_id()?;
+        Ok::<_, chainstate::ChainstateError>((is_ibd, actual_tip))
+    })??;
+
+    // Note:
+    // 1) When chainstate receives multiple blocks in rapid succession, mempool may start lagging
+    // behind it significantly. So a situation is possible when the chainstate is out of ibd
+    // already, but new_tip corresponds to an earlier block, which was obtained when the node
+    // was still in ibd.
+    // 2) If we allowed mempool to handle new_tip events normally while it's lagging, it would lead
+    // to issues. E.g. in the past it was possible for mempool to try connecting transactions of
+    // a past block to some of that block's descendants because of this.
+    // So we bail out if new_tip isn't equal to the actual tip of the chainstate.
+    // (Note that this check doesn't fully prevent mempool from falling out of sync with chainstate.
+    // E.g a new chainstate tip may appear while mempool is in the process of reorging transactions,
+    // which still may cause issues).
+    if is_ibd || new_tip != actual_tip {
+        log::debug!("Not updating mempool: is_ibd = {is_ibd}, new_tip = {new_tip:?}, actual_tip = {actual_tip:?}");
 
         // We still need to update the current tx_verifier tip
         let mut old_transactions = mempool.reset();
         if old_transactions.next().is_some() {
-            log::warn!("Discarding mempool transactions during IBD");
+            log::warn!("Discarding mempool transactions during IBD or because the tip has changed");
         }
         return Ok(());
```

Comment:
Btw, is this needed when the tip is out of sync? Why was it needed in the ibd case?

Reply:
I think in the IBD case, if the IBD correctly behaves as a latch, it should not happen. I just found it a bit risky to go for a full assertion. I would say dropping transactions from mempool is not appropriate when mempool is temporarily out of sync. However, when mempool catches up, it should `reset()` and reverify to see which transactions still apply at the current tip.

Reply:
It turned out that some functional tests do depend on this "mempool.reset" during ibd, so I left this call under "if is_ibd" and wrote a comment.